### PR TITLE
Make crosswalks bidirectional

### PR DESCRIPTION
**Pull request by mdejean**
For issue #813

Problems:

* ~~Currently crashing in map_model::objects::traffic_signals::Stage::max_crosswalk_time failing to find a movement in i.movements~~
* I added PathStep::ContraflowTurn mostly by adding it to the enum and then fixing all the `non-exhaustive pattern` errors, which means I likely missed some `match`es which use `_`.
* I haven't touched pathfinder v2 yet, notably `into_v1_walking` would likely generate TurnIDs which do not correspond to any Turn

other TODOs:

* Degenerate intersections still have two crosswalks (they had four before?). Do they not have `SharedSidewalkCorner`s?
* Once all duplicate crosswalks are not being created, remove `other_crosswalk_ids` and its one reference.

### CODE REVIEW

File: map_model/src/pathfind/v1.rs [hunk 1]
```diff
@@ -675,11 +686,13 @@ fn validate_continuity(map: &Map, steps: &[PathStep]) {
             PathStep::Lane(id) => map.get_l(id).last_pt(),
             PathStep::ContraflowLane(id) => map.get_l(id).first_pt(),
             PathStep::Turn(id) => map.get_t(id).geom.last_pt(),
+            PathStep::ContraflowTurn(id) => map.get_t(id).geom.first_pt(),
         };
         let to = match pair[1] {
             PathStep::Lane(id) => map.get_l(id).first_pt(),
             PathStep::ContraflowLane(id) => map.get_l(id).last_pt(),
             PathStep::Turn(id) => map.get_t(id).geom.first_pt(),
+            PathStep::ContraflowTurn(id) => map.get_t(id).geom.first_pt(),
```
Comment (dabreegster) [inline map_model/src/pathfind/v1.rs:695]: Should be `last_pt`, right? This validation is commented out, could be good to enable it to double-check

File: map_model/src/pathfind/walking.rs [hunk 2]
```diff
@@ -299,21 +299,25 @@ fn make_input_graph(
...
-                / PathStep::Turn(t.id).max_speed_along(max_speed, PathConstraints::Pedestrian, map);
-            if t.turn_type == TurnType::UnmarkedCrossing {
-                // TODO Add to RoutingParams
-                cost = 3.0 * cost;
+                nodes.get(WalkingNode::SidewalkEndpoint(dst.get_directed_parent(), dst.dst_i == t.id.parent));
+            for (step, pair) in [
+                (PathStep::Turn(t.id), (from, to)),
+                (PathStep::ContraflowTurn(t.id), (to, from)),
+            ] {
+                let mut cost = t.geom.length()
```
Comment (dabreegster) [inline map_model/src/pathfind/walking.rs:309]: Nit: I'd leave this outside the loop, just to visually make it more clear that it's the same cost for both directions of the turn

**Review by dabreegster — COMMENTED** [review]
> Currently crashing in map_model::objects::traffic_signals::Stage::max_crosswalk_time failing to find a movement in i.movements

Hmm.. I tried adding `(to, from)` to https://github.com/a-b-street/abstreet/blob/e655e4e1da08f96e1fcd39a6ea2666145166c894/map_model/src/objects/movement.rs#L73, but that didn't appear to help. I wonder if we need to extend the concept of contraflow to `Movement` as well. It's meant to be a less-detailed version of a `Turn` for vehicles -- some straight turn from road1 to 2, doesn't matter exactly which lane. Crosswalks are also encoded by movements for consistency. I wonder if we need to include a direction in https://github.com/a-b-street/abstreet/blob/e655e4e1da08f96e1fcd39a6ea2666145166c894/map_model/src/objects/movement.rs#L21 and duplicate the movements, or... "canonicalize" when we create a crosswalk movement ID and always force the `from` road ID to be `<` the `to` ID. Not sure about the right approach...

> I added PathStep::ContraflowTurn mostly by adding it to the enum and then fixing all the non-exhaustive pattern errors, which means I likely missed some matches which use _.

I wish there was a clippy lint to help with this. I think the workaround is to grep the codebase for `PathStep::Contraflow` and skim through call-sites.

> I haven't touched pathfinder v2 yet, notably into_v1_walking would likely generate TurnIDs which do not correspond to any Turn

https://github.com/a-b-street/abstreet/blob/e655e4e1da08f96e1fcd39a6ea2666145166c894/map_model/src/pathfind/v2.rs#L243
So I think two things have to happen here... first we need to generate the correct `TurnID`. Either the movement IDs that get created in the first place always need to match up with the arbitrary from,to order in the `TurnID`. Or when we create the crosswalk turns in the first place, force the IDs to follow `from < to` ordering or whatever.

Then we have to pick between following the movement normally or contraflow. We might be forced to look at the previous step and figure out where we ended. Basically following the logic from `validate_continuity`

> Degenerate intersections still have two crosswalks (they had four before?). Do they not have SharedSidewalkCorners?

Screenshot from 2021-12-03 06-50-26
(You can hover on a sidewalk and press **z** in dev mode to browse turns)
So for this case, we still do have shared sidewalk corners... 4 of them, 2 on each side of the intersection, and a pair going forwards and then backwards.

The reason the crosswalks have 4 originally is similar. Actually just seeing it is easier than explaining:
Screenshot from 2021-12-03 06-50-26
With your change, this now becomes
Screenshot from 2021-12-03 06-53-23
We could also reconsider this choice. Maybe a pedestrian has to cross a shared sidewalk corner to get to the other sidewalk, then cross just a single crosswalk (normally or contraflow). There would probably be a visual "glitch" and extra distance on their path as they slightly double back. This already happens at normal intersections, because the crosswalk is only connected to one end of one sidewalk, even though kind of logically it's at a corner / diagonal. A video maybe illustrates:
screencast
They just want to walk east. But because the crosswalk is connected to the north/south sidewalks, they have to follow a SharedSidewalkCorner on both ends. This has always annoyed me.

File: map_model/src/pathfind/walking.rs (hunk 2 above)
Comment (mdejean) [reply map_model/src/pathfind/walking.rs]: After rereading I eliminated the loop and replaced it with two calls to add_edge

**Comment by mdejean** [comment]
The crash was caused by edit_movement https://github.com/a-b-street/abstreet/blob/master/map_model/src/objects/traffic_signals.rs#L331

**Comment by mdejean** [comment]
>We could also reconsider this choice. Maybe a pedestrian has to cross a shared sidewalk corner to get to the other sidewalk, then cross just a single crosswalk (normally or contraflow). There would probably be a visual "glitch" and extra distance on their path as they slightly double back. 

I don't think that's an issue, though. The crosswalk is where it is, even if they have to walk slightly further to it. I don't think this is the right place to solve the problem with crosswalks being connected to the edge of the intersection while the sidewalk corner is logically inside it.

**Comment by mdejean** [comment]
* Currently crashing in sim::mechanics::walking::WalkingSimState::spawn_ped on calling as_lane on something that's not a lane
* Changed to make_walking_turns_v2 and make_shared_sidewalk_corner using Ring::get_shorter_slice_btwn which is currently not quite working(?)
* Need to make make_walking_turns_v2 handle footways

**Comment by dabreegster** [comment]
Nice to see the progress!

> Changed to make_walking_turns_v2 and make_shared_sidewalk_corner using Ring::get_shorter_slice_btwn which is currently not quite working(well I think it is, but it isn't?)

I don't remember why this is so hard, but it's hard. :(  Right now I'm seeing the montlake map crash in `make_walking_turns`, and `arboretum` doesn't crash but regresses:
Screenshot from 2021-12-04 21-44-09

If it's helpful, I made two recent attempts to simplify `make_shared_sidewalk_corner`:
https://github.com/a-b-street/abstreet/tree/simplify_sidewalk_corners
https://github.com/a-b-street/abstreet/tree/shorter_slice_btwn

I believe one of them nearly worked, but visually, some of the output got worse.

**Comment by mdejean** [comment]
Very much broke sharedsidewalkcorner rendering but the geometry looks okay here.
image

I'm very confused by what `calculate_corners` is trying to do so I'll probably stab that next.

**Comment by mdejean** [comment]
Got down a rabbit hole trying to do the sharedsidewalkcorner thing, so i'm going to back that out.

**Comment by mdejean** [comment]
Now generating and rendering `SharedSidewalkCorner`s correctly regardless of driving side
image

* Footways are making a crosswalk instead of a sharedsidewalkcorner. Need to figure out how to handle footways, especially intersections which have both footways and roads.
image

**Comment by mdejean** [comment]
Behold!
image

TODO:

* Footway turns
* Pedestrian waiting-to-turn arrows
* Check pathfinding is actually working
* Crosswalks still being duplicated in some cases:
image

**Comment by mdejean** [comment]
footway turns ✓
image

pathfinding working ✓
image

**Comment by mdejean** [comment]
I'm going to go ahead with this at this point

Right now outstanding issues are:
* Pedestrian waiting-to-turn arrows
I'll fold this into the curvy-turn-arrows change to relative angles

* Crosswalks still being duplicated in some cases
This results from the check https://github.com/mdejean/abstreet/blob/contraflow_turn/map_model/src/make/walking_turns.rs#L108

which really should be something like if there is already a crosswalk in `result` which connects l1 to l2 through one or more `sharedsidewalkcorner`s, which is not an easy condition to write.

* Pedestrians walking around corners to get to the crosswalk.

I'll address this one separately. I'm trying to think of a way to squash together the sharedsidewalkcorner turn and crosswalk turns to eliminate any duplicated geometry.

**Comment by mdejean** [comment]
I'll squash the commits today. There are a couple of unrelated things in there that I'll pull out.

**Comment by mdejean** [comment]
that was relatively painless

File: map_model/src/make/walking_turns.rs [hunk 3]
```diff
@@ -1,235 +1,21 @@
...
-                // side
-                result.extend(
-                    make_crosswalks(i.id, l1, l2, driving_side)
-                        .into_iter()
-                        .flatten(),
-                );
-            } else {
-                // We may need to add a crosswalk over this intermediate road that has no
-                // sidewalks at all. There might be a few in the way -- think highway onramps.
-                // TODO Refactor and loop until we find something to connect it to?
```
Comment (dabreegster) [inline map_model/src/make/walking_turns.rs:137]: Good riddance, first version of this code!

File: map_model/src/make/walking_turns.rs [hunk 4]
```diff
@@ -1,235 +1,21 @@
...
-                    turn.turn_type = TurnType::UnmarkedCrossing;
-                }
-            }
-        }
-    }
-
-    input
-}
-
 /// A complete rewrite of make_walking_turns, which looks at all sidewalks (or lack thereof) in
```
Comment (dabreegster) [inline map_model/src/make/walking_turns.rs:10]: You could remove the "rewrite" part now

File: map_model/src/map.rs [hunk 5]
```diff
@@ -320,39 +320,35 @@ impl Map {
...
+        let turns: Vec<&Turn> = self
             .get_i(lane.dst_i)
             .turns
             .iter()
-            .filter(|t| t.id.src == l)
+            .chain(
+                self.get_i(lane.src_i)
+                    .turns
+                    .iter()
+                    .take_while(|_| lane.is_walkable()),
```
Comment (dabreegster) [inline map_model/src/map.rs:331]: This relies on all of the walking turns being placed before the vehicle turns, right? This could be brittle; `filter` might be safer. If we do want the short-circuiting here, then we should add a docstring on `turns: Vec<Turn>` then

File: map_model/src/pathfind/walking.rs [hunk 6]
```diff
@@ -298,22 +298,26 @@ fn make_input_graph(
         if t.between_sidewalks() {
             let src = map.get_l(t.id.src);
             let dst = map.get_l(t.id.dst);
```
Comment (dabreegster) [inline map_model/src/pathfind/walking.rs:300]: `MovementID` is now an unused import in this file, please fix

File: map_model/src/objects/turn.rs [hunk 7]
```diff
@@ -1,4 +1,4 @@
-use std::collections::BTreeSet;
+
```
Comment (dabreegster) [inline map_model/src/objects/turn.rs:1]: Stray whitespace detected by cargo fmt

**Review by dabreegster — COMMENTED** [review]
`outgoing_lanes` in `objects/road.rs` becomes unused, we can remove it.

The code looks fine to me! Lots of it is much simpler now. I didn't follow the geometry in detail, but I'll trust the results.

I'll next regenerate all maps with this change and run through all the screenshot diff tests and make sure nothing regresses really badly. That'll also entail cleaning up the baked-in traffic signal data as I mentioned. Should hopefully get to that by morning. Thanks for the monumental effort behind this PR!

File: map_model/src/map.rs (hunk 5 above)
Comment (mdejean) [reply map_model/src/map.rs:331]: No, this is the way I could find to `chain` all the turns from the source (i.e. contraflow) intersection only if `lane.is_walkable()`

**Comment by dabreegster** [comment]
Alright, finally figured out how to sanely work on this PR -- for my future reference, the github CLI is way easier than trying to [follow these docs](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/committing-changes-to-a-pull-request-branch-created-from-a-fork). All I did was `gh pr checkout 816`, then I have a local branch that's synced up with yours.

I rebased and dealt with a tiny merge conflict. Now I'll fix up the traffic signals and regenerate all maps!

**Comment by dabreegster** [comment]
The only problem I spot is duplicate crosswalks in some cases.

The montlake map (https://www.openstreetmap.org/node/53198127):
Screenshot from 2021-12-16 10-01-07

gb/poundbury (https://www.openstreetmap.org/node/907245890):
Screenshot from 2021-12-16 10-02-03
Many roundabouts here show the same thing.

You identified the reason earlier:

> This results from the check https://github.com/mdejean/abstreet/blob/contraflow_turn/map_model/src/make/walking_turns.rs#L108 which really should be something like if there is already a crosswalk in result which connects l1 to l2 through one or more sharedsidewalkcorners, which is not an easy condition to write.

I'll make an attempt later today to implement this. It's a bit of a large regression to leave hanging for a while, but also the amount of cleanup in this PR is very tempting to merge in.

**Comment by mdejean** [comment]
>>  This results from the check https://github.com/mdejean/abstreet/blob/contraflow_turn/map_model/src/make/walking_turns.rs#L108 which really should be something like if there is already a crosswalk in result which connects l1 to l2 through one or more sharedsidewalkcorners, which is not an easy condition to write.
> 
> I'll make an attempt later today to implement this. It's a bit of a large regression to leave hanging for a while, but also the amount of cleanup in this PR is very tempting to merge in.

Actually this condition isn't sufficient because the connecting sharedsidewalkcorner might not have been made yet. So I think it'll need to be a fixup at the end removing connected crosswalks.

**Comment by mdejean** [comment]
This happens only when there are exactly two crosswalks. It can happen with any number of sharedsidewalkcorners due to footways. It can happen with any number of roads without sidewalks (including zero like a degenerate intersection or a footway intersection).

**Comment by mdejean** [comment]
Is it even possible for an intersection to have exactly two crosswalks in any case except this?

**Comment by dabreegster** [comment]
Ahh, that makes sense! That commit fixes the two problems above, and I'm not spotting any more issues. So in that case, I'll start regenerating everything, upload the new data, then merge. Sound like a plan?

**Comment by mdejean** [comment]
Sounds good.

**Comment by michaelkirk** [comment]
It was fun watching you methodically untie this gordian knot these last two weeks @mdejean.

And I love to see fixes which delete more code than they add! (even before all the traffic signal data was deleted).

Good work! 🎉

**Comment by mdejean** [comment]
I'm getting a panic sometimes when watching zoomed in intersections at

https://github.com/mdejean/abstreet/blob/contraflow_turn/sim/src/mechanics/walking.rs#L693

**Comment by dabreegster** [comment]
> I'm getting a panic sometimes when watching zoomed in intersections at

Hmm... know which map or intersection? I'd guess it's when turns or crosswalks have really short geometry. In `make_shared_sidewalk_corner`, the fallback could collapse to a single point, `// This may produce a polyline with two identical points. Nothing better to do here.`. That might break lots of other `geom` stuff downstream; invalid polylines with 0 length or duplicate adjacent points are usually only constructed temporarily for measuring length or trying to generate a weird outline shape. Sometimes a corner legitimately has no length; maybe we'll be forced to have `Option<PolyLine>`? I'd like to reproduce the bug first and understand that place.

Another crash: while running the screenshot test in `pl/krakow`, I got a `called `Option::unwrap()` on a `None` value', geom/src/line.rs:24:29`.

Screenshot test in `us/phoenix/maps/tempe.bin` also showed a few things I'd like to check into. 
Screenshot from 2021-12-17 08-20-17
At dead-ends, we now have a crosswalk instead of a corner wrapping around:
Screenshot from 2021-12-17 08-21-29
Originally I was trying to model little cul-de-sacs, where the sidewalk often does wrap around the whole shape.

I'll add docs on how to run the [screenshot test](https://a-b-street.github.io/docs/tech/dev/testing.html#screenshot-diffs) in a moment. I can dig into these issues over the weekend too.

**Comment by mdejean** [comment]
So I fixed the deadends not being sharedsidewalkcorners and in the process of investigating the deadend sidewalk not ending up square cleaned up make_shared_sidewalk_corner a lot.

**Comment by mdejean** [comment]
I found another issue: a degenerate intersection on a street with only one sidewalk generates both a crosswalk and a SharedSidewalkCorner
image
image

**Comment by mdejean** [comment]
> I'll add docs on how to run the screenshot test in a moment. I can dig into these issues over the weekend too.

My monitor is 1280x1024 so I can't do that :roll_eyes:

**Comment by dabreegster** [comment]
Noticed another issue with the dead-ends: the curb's position (from `calculate_corner_curbs`) is inverted:
Screenshot from 2021-12-18 19-54-51

> My monitor is 1280x1024 so I can't do that

It's a brittle test, sadly... but you _could_ run the test on the master branch, `confirm_screencaps.sh` to update your local copy of the screenshots, then run the usual test on this branch.

**Comment by mdejean** [comment]
I patched up polyline::dist_along so it calls line::dist_along instead of line::must_dist_along and passes along an error, that made the `expect` in get_draw_ped catch it.

i put in some info and got: `thread 'main' panicked at 'at 00:01:00.0, Pedestrian #404's position percent: 0.0000110998878911323 DistanceInterval { start: Distance(9.7055), end: Distance(0.0) } of Turn(TurnID { parent: IntersectionID(91), src: LaneID { road: RoadID(323), offset: 4 }, dst: LaneID { road: RoadID(323), offset: 0 } }) (9.7055m long) is broken`

which is the crosswalk here
image

I then watched several pedestrians use this crosswalk in both directions without crashing and got.

`'at 00:03:00.0, Pedestrian #2794's position percent: 0.000009255488504683279 DistanceInterval { start: Distance(9.7056), end: Distance(0.0) } of Turn(TurnID { parent: IntersectionID(154), src: LaneID { road: RoadID(132), offset: 0 }, dst: LaneID { road: RoadID(132), offset: 4 } }) (9.7056m long) is broken: ', sim\src\mechanics\walking.rs:703:22`

which is another nearby intersection, so i'm guessing this is some interaction of floating point stuff with polyline::dist_along and dist_int.lerp where a pedestrian right at the beginning of a contraflow turn can end up calling dist_along with very slightly less than the length, but floating point errors add up in polyline::dist_along and there's a little bit left over. I did manage to get this to not happen by changing the check 
        if dist_along == self.length() {
            return Ok((self.last_pt(), self.last_line().angle()));
        }
to
        if dist_along > self.length() - EPSILON_DIST {
            return Ok((self.last_pt(), self.last_line().angle()));
        }

IMO, up to you about how to deal with this.

**Comment by dabreegster** [comment]
I rebased against master and fixed a crash in Krakow, caused by forcibly creating a polyline with 2 equal points. I just squinted carefully at the screenshot test for the 5 maps, and I'm happy with the diffs. I'm ready to go ahead and get this merged in. Any objections?

> this is some interaction of floating point stuff with polyline::dist_along and dist_int.lerp where a pedestrian right at the beginning of a contraflow turn can end up calling dist_along with very slightly less than the length, but floating point errors add up in polyline::dist_along and there's a little bit left over

Doing the epsilon check seems sensible to me, yeah. I've tried to keep those from proliferating everywhere, but they're inevitable, I think. I couldn't reproduce this one in any of the maps I've been testing; where is Davidson ave?

**Comment by mdejean** [comment]
That's on the fordham map. This crash should happen on every map, but I haven't been able to trigger it anywhere else. There's nothing special about the geometry at that intersection. AFAIK it happens with every contraflow turn with multiple line segments (e.g. every crosswalk), but the simulation needs to step just right for it to happen.

**Comment by dabreegster** [comment]
Let's just hope all the floating point stuff is deterministic across platforms; I want to reproduce the crash you see and try that fix.

> That's on the fordham map

Could we check in this importer config so I can follow more easily? I assume this is https://github.com/mdejean/abstreet/blob/fordham/importer/config/us/nyc/fordham.poly?

**Comment by mdejean** [comment]
> Could we check in this importer config so I can follow more easily? I assume this is https://github.com/mdejean/abstreet/blob/fordham/importer/config/us/nyc/fordham.poly?

Yes.

**Comment by mdejean** [comment]
Put it on random trips and it should crash after a minute or three

**Comment by dabreegster** [comment]
> Put it on random trips and it should crash after a minute or three

Nope, all good.. both `home_to_work` and the "random unrealistic." But my intersection IDs and screenshots don't match up with yours:
Screenshot from 2021-12-20 21-01-28
I don't see any parking lanes on the map at all. Are you using special importer config somehow?

**Comment by mdejean** [comment]
You might need to delete/redownload data/input/us/nyc/osm/new-york-latest.osm.pbf since I've added many of those in the last few weeks.

**Comment by dabreegster** [comment]
Nice, it's awesome to see lanes filled out so thoroughly in a concentrated area! Unfortunately, the IDs still don't match up, and I can't reproduce the crash.

I propose:
1) I'll regenerate all maps and merge this PR
2) You could send the binary map file, so I can try to repro the crash
3) We make the epsilon fix separately. (It'll also require a full map rebuild, since any fix deep in `geom` potentially affects everything

**Comment by dabreegster** [comment]
That beautiful moment of pushing data to S3 when turning back would be way too much hassle... Alright, thank you so much for powering through this! Merging momentarily when uploading is done

**Comment by mdejean** [comment]
IDs not matching up might have been that I was using `--consolidate-all-intersections` though the crash happens either way

Here is abstreet/data/system/us/nyc/maps/fordham.bin
[fordham.bin.zip](https://github.com/a-b-street/abstreet/files/7748529/fordham.bin.zip)

**Comment by Robinlovelace** [comment]
Amazing work @mdejean, this is an epic PR that hugely improves how A/B Street handles walking, as far as I can tell.

**Comment by Robinlovelace** [comment]
Would be good to have a before/after GIF to show the consequences... This is newsworthy!

**Comment by mdejean** [comment]
> Would be good to have a before/after GIF to show the consequences... This is newsworthy!

There isn't a major visual impact.

* Crosswalks may have changed a little bit due to the change to the new walking turns generator
* Crosswalks are no longer double rendered on the current stage when using Yuwen's style.
